library/src/layout/par.rs

```diff
@@ -864,14 +876,31 @@ fn linebreak_optimized<'a>(vt: &Vt, p: &'a Preparation<'a>, width: Abs) -> Vec<L
             // Determine how much the line's spaces would need to be stretched
             // to make it the desired width.
             let delta = width - attempt.width;
-            let mut ratio = delta / attempt.stretch();
+            // Determine how much stretch are permitted.
+            let adjust = if delta >= Abs::zero() {
+                attempt.stretchability()
+            } else {
+                attempt.shrinkability()
+            };
+            // Ideally, the ratio should between -1.0 and 1.0, but sometimes a value above 1.0
+            // is possible, in which case the line is underfull.
+            let mut ratio = delta / adjust;
+            if ratio.is_nan() {
+                // The line is not stretchable, but it just fits.
+                // This often happens with monospace fonts and CJK texts.
+                ratio = 0.0;
+            }
             if ratio.is_infinite() {
+                // The line's not stretchable, we calculate the ratio in another way...
                 ratio = delta / (em / 2.0);
+                // ...and because it is underfull/overfull, make sure the ratio is at least 1.0.
+                if ratio > 0.0 {
+                    ratio += 1.0;
+                } else {
+                    ratio -= 1.0;
+                }
```

Comment:
I don't understand this part. But to be honest, I also don't understand anymore what I did above with `delta / (em / 2.0)`.

Reply:
I don't understand why you use `delta / (em / 2.0)` too 😂 
But I think the formula can be chosen relatively randomly and that's fine in case it works. After all, Knuth also admitted his cost formula is quite arbitrary.

Your formula do have a problem though. It cannot ensure underfull and overfull line get the ratio out the range of $[-1.0, 1.0]$. So I add these code as a patch.

The reason why the range is important is partly explained in last comment: https://github.com/typst/typst/pull/701#discussion_r1164860542